### PR TITLE
🛠️ Backend threshold

### PR DESCRIPTION
**Pull request by bodmer02**
# Pull Request

## Changes

- Add Jacoco branch coverage check
- Add excludes
- Add tests

## Reference

Issue: #250

## Checklist

- [x] ~Updated documentation~
- [x] Added automated tests (unit/integration/component)
- [x] ~Complied with UI/UX design (if UI change was made)~
- [x] Completed a portion of the acceptance criteria for the issue.

## Summary by CodeRabbit

* **Tests**
  * Reorganized and expanded unit tests into nested suites, added null/edge-case tests, introduced new unit tests for logging and auth utilities, and added an integration test verifying Cache-Control behavior.

* **Chores**
  * Enforced a branch coverage threshold (0.80) and updated coverage reporting to exclude generated/infrastructure code.
  * Added and applied a marker annotation to exempt selected classes/methods from coverage calculations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/KeycloakRolesAuthoritiesConverterTest.java [hunk 1]
```diff
@@ -35,67 +39,79 @@ void setUp() {
...
+            final Jwt jwt = mock(Jwt.class);
+            when(jwt.getClaimAsMap(RESOURCE_ACCESS_CLAIM)).thenReturn(claims);
+
+            // Call
+            final Collection<GrantedAuthority> authorities = converter.convert(jwt);
+
+            // Assert
+            assert authorities != null;
+            assertTrue(authorities.isEmpty());
+        }
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/KeycloakRolesAuthoritiesConverterTest.java]: Changed "assert authorities != null" to "Assertions.assertNotNull(authorities)"
Changed mock to return resourceAccessClaim instead of the entire claims map.

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/RequestResponseLoggingFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        void testLoggingEnabledForChangingMode() throws IOException, ServletException {
+            when(securityProperties.getLoggingMode()).thenReturn(RequestResponseLoggingFilter.LoggingMode.CHANGING);
+            when(mockRequest.getMethod()).thenReturn(HttpMethod.POST.name());
+            when(mockRequest.getRequestURI()).thenReturn("/test");
+            when(mockResponse.getStatus()).thenReturn(200);
+            filter.doFilterInternal(mockRequest, mockResponse, mockFilterChain);
+            verify(mockFilterChain).doFilter(mockRequest, mockResponse);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/RequestResponseLoggingFilterTest.java]: Changed test names

File: stadtbezirksbudget-backend/pom.xml [hunk 3]
```diff
@@ -371,6 +371,44 @@
...
+                        <goals>
+                            <goal>report</goal>
+                        </goals>
+                        <configuration>
+                            <excludes>
+                                <exclude>*@de.muenchen.stadtbezirksbudget.backend.common.ExcludedFromGeneratedCoverage*</exclude>
+                                <exclude>**/*MapperImpl.class</exclude>
+                            </excludes>
+                        </configuration>
+                    </execution>
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/pom.xml]: It is also working with CLASS retention

File: stadtbezirksbudget-backend/src/main/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/SecurityConfiguration.java [hunk 4]
```diff
@@ -1,5 +1,6 @@
 package de.muenchen.stadtbezirksbudget.backend.configuration.security;
 
+import de.muenchen.stadtbezirksbudget.backend.common.ExcludedFromGeneratedCoverage;
 import java.util.Optional;
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/main/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/SecurityConfiguration.java]: Added "for branch coverage" in the comment

File: stadtbezirksbudget-backend/src/main/java/de/muenchen/stadtbezirksbudget/backend/kafka/KafkaConsumerService.java [hunk 5]
```diff
@@ -22,6 +23,8 @@
  */
 @Service
 @Slf4j
+@ExcludedFromGeneratedCoverage
+//Excluded because no further implementation exists, tests will be implemented later. TODO: Add tests and remove exclusion to reach branch coverage
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/main/java/de/muenchen/stadtbezirksbudget/backend/kafka/KafkaConsumerService.java]: Created Issue and added Number to TODO

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/RequestResponseLoggingFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        @Test
+        void testFilterDelegatesWithoutErrorForChangingMode() throws IOException, ServletException {
+            when(securityProperties.getLoggingMode()).thenReturn(RequestResponseLoggingFilter.LoggingMode.CHANGING);
+            when(mockRequest.getMethod()).thenReturn(HttpMethod.POST.name());
+            when(mockRequest.getRequestURI()).thenReturn("/test");
+            when(mockResponse.getStatus()).thenReturn(200);
+            filter.doFilterInternal(mockRequest, mockResponse, mockFilterChain);
+            verify(mockFilterChain).doFilter(mockRequest, mockResponse);
+        }
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/RequestResponseLoggingFilterTest.java]: Added additional test with different status value

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/UserInfoAuthoritiesConverterTest.java [hunk 7]
```diff
@@ -47,63 +52,139 @@ void setUp() {
...
+
+        @Test
+        void testConvertWithCollectionAuthorities() {
+            final Jwt jwt = mock(Jwt.class);
+            when(jwt.getSubject()).thenReturn(TEST_SUBJECT);
+            when(jwt.getTokenValue()).thenReturn(TEST_TOKEN_VALUE);
+            final Map<String, Object> responseMap = new HashMap<>();
+            responseMap.put(AUTHORITIES, List.of("ROLE_USER", "ROLE_ADMIN")); // Simulate returning a collection
+            when(restTemplate.exchange(eq(USER_INFO_URI), eq(HttpMethod.GET), any(), eq(Map.class)))
+                    .thenReturn(new ResponseEntity<>(responseMap, HttpStatus.OK));
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/UserInfoAuthoritiesConverterTest.java]: Implemented "when(restTemplate.exchange(eq(USER_INFO_URI), eq(HttpMethod.GET), any(HttpEntity.class), eq(Map.class)))" for matcher type-safety

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/UserInfoAuthoritiesConverterTest.java [hunk 8]
```diff
@@ -47,63 +52,139 @@ void setUp() {
...
+        @Test
+        void testCacheHitWithNullValue() {
+            final Jwt jwt = mock(Jwt.class);
+            when(jwt.getSubject()).thenReturn("abc");
+            final Cache.ValueWrapper wrapper = mock(Cache.ValueWrapper.class);
+            when(wrapper.get()).thenReturn(null);
+            when(cache.get("abc")).thenReturn(wrapper);
+            final Collection<GrantedAuthority> result = converter.convert(jwt);
+            assertNull(result);
+        }
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/security/UserInfoAuthoritiesConverterTest.java]: We are only testing the code, we do not edit the template code.

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/CacheControlFilterTest.java [hunk 9]
```diff
@@ -35,20 +48,60 @@ class CacheControlFilterTest {
...
-        assertEquals(HttpStatus.OK, response.getStatusCode());
-        assertTrue(response.getHeaders().containsKey(HttpHeaders.CACHE_CONTROL));
-        assertEquals(EXPECTED_CACHE_CONTROL_HEADER_VALUES, response.getHeaders().getCacheControl());
+    @BeforeEach
+    void setUp() {
+        filter = new CacheControlFilter();
+        request = mock(HttpServletRequest.class);
+        response = mock(HttpServletResponse.class);
+        filterChain = mock(FilterChain.class);
+    }
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/CacheControlFilterTest.java]: Moved the integration tests into a separated class.

File: stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/CacheControlFilterIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        classes = { StadtbezirksbudgetBackend.class },
+        webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT
+)
+@ActiveProfiles(profiles = { SPRING_TEST_PROFILE, SPRING_NO_SECURITY_PROFILE })
+class CacheControlFilterIntegrationTest {
+    @Container
+    @ServiceConnection
+    private static final PostgreSQLContainer<?> POSTGRE_SQL_CONTAINER = new PostgreSQLContainer<>(
+            DockerImageName.parse(TestConstants.TESTCONTAINERS_POSTGRES_IMAGE));
+
```
[the remark this answers is not in the document]
Comment (bodmer02) [reply stadtbezirksbudget-backend/src/test/java/de/muenchen/stadtbezirksbudget/backend/configuration/filter/CacheControlFilterIntegrationTest.java]: Added  "SuppressWarnings("unused")"

**Review by danielgully — APPROVED** [review]
